### PR TITLE
Adds the new Shrine to flutter_gallery

### PR DESCRIPTION
**Pull request by redbrogdon**
This is a redo of PR #22981, which ended up with a bunch of repeated commits in it due to an unfortunate merge/rebase.

Most of this code has already been reviewed by @HansMuller as it made its way into the material codelabs:

https://github.com/material-components/material-components-flutter-codelabs

The only significant changes beyond moving in the previously reviewed code are:

* A new "exit" button on the login screen so users (and tests) can back out of the demo.
* A ScopedModel wrapping the gallery app.
* New references to `scoped_model` and `shrine_images` in the pubspec.
  - The latter contains image assets so they're kept out of the main repo.
* Removal of some unneeded code that used to launch the old demo.
* A new `toString` method for `TweenSequence`. The lack of one was causing some of the tests to fail.

### CODE REVIEW

**Comment by willlarche** [comment]
🎉

**Comment by RedBrogdon** [comment]
Fixes #22722.

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+    if (_isOpen) {
+      _controller.reverse();
+    }
+  }
+
+  // Changes the padding between the start edge of the Material and the cart icon
+  // based on the number of products in the cart (padding increases when > 0
+  // products.)
+  EdgeInsetsDirectional _cartPaddingFor(int numProducts) {
+    if (numProducts == 0) {
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Typically
```
return (numProducts == 0)
  ? const EdgeInsetsDirectional.only(start: 20.0, end: 8.0);
  : const EdgeInsetsDirectional.only(start: 32.0, end: 8.0);
```

Maybe a little clearer as:
```
return EdgeInsetsDirectional.only(
  start: numProducts == 0 ? 20.0 : 32.0
  end: 8.0
);
```

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+  }
+
+  bool get _cartIsVisible => _thumbnailOpacityAnimation.value == 0.0;
+
+  Widget _buildThumbnails(int numProducts) {
+    return ExcludeSemantics(
+      child: Opacity(
+        opacity: _thumbnailOpacityAnimation.value,
+        child: Column(children: <Widget>[
+          Row(children: <Widget>[
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Typically:
```
child: Column(
  children: <Widget>[
    Row(
      children: <Widget>[
        ...
      ],
    ),
    ...
  ],
),
```

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+          animationDuration: const Duration(milliseconds: 0),
+          shape: BeveledRectangleBorder(
+            borderRadius: BorderRadius.only(
+              topLeft: Radius.circular(_shapeAnimation.value),
+            ),
+          ),
+          elevation: 4.0,
+          color: kShrinePink50,
+          child: _cartIsVisible
+              ? _buildShoppingCartPage()
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+      // Maximum of 99 so padding doesn't get messy.
+      final int displayedOverflowProducts = numOverflowProducts <= 99 ? numOverflowProducts : 99;
+      return Container(
+        child: Text(
+          '+$displayedOverflowProducts',
+          style: Theme.of(context).primaryTextTheme.button,
+        ),
+      );
+    } else {
+      return Container(); // build() can never return null.
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: return const SizedBox()

This method might be easier to read if the first line was:
```
if (model.productsInCart.length <= 3)
  return const SizedBox();
// no else clause
...
```

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return ScopedModelDescendant<AppStateModel>(
+      builder: (BuildContext builder, Widget child, AppStateModel model) =>
+          _buildOverflow(model, context),
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: No line break

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+        ),
+      ),
+    );
+  }
+}
+
+// _ListModel manipulates an internal list and an AnimatedList
+class _ListModel {
+  _ListModel(
+      {@required this.listKey,
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Typically
```
_ListModel({
  @required this.listKey,
  @required this.removedItemBuilder,
  Iterable<int> initialItems,
}) {
  ...
}
```

File: examples/flutter_gallery/lib/demo/shrine/home.dart [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import 'package:flutter_gallery/demo/shrine/supplemental/asymmetric_view.dart';
+
+class ProductPage extends StatelessWidget {
+  const ProductPage({this.category = Category.all});
+
+  final Category category;
+
+  @override
+  Widget build(BuildContext context) {
+    return ScopedModelDescendant<AppStateModel>(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/home.dart]:
```
Widget build(BuildContext context) {
  return ScopedModelDescendant<AppStateModel>(
    builder: (BuildContext context, Widget child, AppStateModel model) {
      return AsymmetricView(products: model.getProducts());
    },
  );
}
```

File: examples/flutter_gallery/lib/demo/shrine/home.dart [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        builder: (BuildContext context, Widget child, AppStateModel model) {
+      return AsymmetricView(products: model.getProducts());
+    });
+  }
+}
+
+class HomePage extends StatelessWidget {
+  const HomePage({Key key, this.expandingBottomSheet, this.backdrop})
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/home.dart]: Constructor named parameter formatting: `const HomePage({ Key key, ... }) : super(key: key);

File: examples/flutter_gallery/lib/demo/shrine/home.dart [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      : super(key: key);
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/home.dart]: No line break

File: examples/flutter_gallery/lib/demo/shrine/login.dart [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                  labelText: 'Password',
+                ),
+              ),
+            ),
+            ButtonBar(
+              children: <Widget>[
+                Tooltip(
+                  message: 'Back',
+                  child: FlatButton(
+                    child: const Text(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/login.dart]: For Text constants with no parameters we typically just put the constructor one line: `const Text('foo'),`

File: examples/flutter_gallery/lib/demo/shrine/login.dart [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+          ],
+        ),
+      ),
+    );
+  }
+}
+
+class PrimaryColorOverride extends StatelessWidget {
+  const PrimaryColorOverride({Key key, this.color, this.child})
+      : super(key: key);
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/login.dart]: no line break

File: examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  Map<int, int> get productsInCart => Map<int, int>.from(_productsInCart);
+
+  // Total number of items in the cart.
+  int get totalCartQuantity => _productsInCart.values.fold(0, (int v, int e) => v + e);
+
+  Category get selectedCategory => _selectedCategory;
+
+  // Totaled prices of the items in the cart.
+  double get subtotalCost => _productsInCart.keys
+      .map((int id) => _availableProducts[id].price * _productsInCart[id])
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      .fold(0.0, (double sum, int e) => sum + e);
+
+  // Total shipping cost for the items in the cart.
+  double get shippingCost =>
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart]: Only use `=>` for one liners that return a value

File: examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+  Map<int, int> get productsInCart => Map<int, int>.from(_productsInCart);
+
+  // Total number of items in the cart.
+  int get totalCartQuantity => _productsInCart.values.fold(0, (int v, int e) => v + e);
+
+  Category get selectedCategory => _selectedCategory;
+
+  // Totaled prices of the items in the cart.
+  double get subtotalCost => _productsInCart.keys
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart]: Only use `=>` for one liners that return a value

File: examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+  List<Product> getProducts() {
+    if (_availableProducts == null) {
+      return <Product>[];
+    }
+
+    if (_selectedCategory == Category.all) {
+      return List<Product>.from(_availableProducts);
+    } else {
+      return _availableProducts
+          .where((Product p) => p.category == _selectedCategory)
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/model/app_state_model.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine/shopping_cart.dart [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+class ShoppingCartSummary extends StatelessWidget {
+  const ShoppingCartSummary({this.model});
+
+  final AppStateModel model;
+
+  @override
+  Widget build(BuildContext context) {
+    final TextStyle smallAmountStyle = Theme.of(context).textTheme.body1.copyWith(color: kShrineBrown600);
+    final TextStyle largeAmountStyle = Theme.of(context).textTheme.display1;
+    final NumberFormat formatter = NumberFormat.simpleCurrency(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/shopping_cart.dart]: Typically
```
final NumberFormat formatter = NumberFormat.simpleCurrency(
  decimalDigits: 2,
  locale: Localizations.localeOf(context).toString(),
);
```

File: examples/flutter_gallery/lib/demo/shrine/shopping_cart.dart [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    this.onPressed,
+  });
+
+  final Product product;
+  final int quantity;
+  final VoidCallback onPressed;
+
+  @override
+  Widget build(BuildContext context) {
+    final NumberFormat formatter = NumberFormat.simpleCurrency(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/shopping_cart.dart]: Typically
```
final NumberFormat formatter = NumberFormat.simpleCurrency(
  decimalDigits: 0,
  locale: Localizations.localeOf(context).toString(),
);
```

File: examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    return List<Container>.generate(_listItemCount(products.length), (int index) {
+      double width = .59 * MediaQuery.of(context).size.width;
+      Widget column;
+      if (index % 2 == 0) {
+        /// Even cases
+        final int bottom = _evenCasesIndex(index);
+        column = TwoProductCardColumn(
+          bottom: products[bottom],
+          top: products.length - 1 >= bottom + 1
+              ? products[bottom + 1]
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+        child: Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 16.0),
+          child: column,
+        ),
+      );
+    }).toList();
+  }
+
+  int _evenCasesIndex(int input) {
+    /// The operator ~/ is a cool one. It's the truncating division operator. It
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart]: Wow

File: examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    return input ~/ 2 * 3;
+  }
+
+  int _oddCasesIndex(int input) {
+    assert(input > 0);
+    return (input / 2).ceil() * 3 - 1;
+  }
+
+  int _listItemCount(int totalItems) {
+    if (totalItems % 3 == 0) {
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/asymmetric_view.dart]: Typically
```
return totalItems % 3 == 0
  ? totalItems ~/ 3 * 2
  : (totalItems / 3).ceil() * 2 - 1;
```

File: examples/flutter_gallery/lib/demo/shrine/supplemental/cut_corners_border.dart [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+import 'package:flutter/widgets.dart';
+
+class CutCornersBorder extends OutlineInputBorder {
+  const CutCornersBorder({
+    BorderSide borderSide = BorderSide.none,
+    BorderRadius borderRadius = const BorderRadius.all(Radius.circular(2.0)),
+    this.cut = 7.0,
+    double gapPadding = 2.0,
+  }) : super(
+            borderSide: borderSide,
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/cut_corners_border.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine/supplemental/cut_corners_border.dart [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+    BorderSide borderSide,
+    BorderRadius borderRadius,
+    double gapPadding,
+    double cut,
+  }) {
+    return CutCornersBorder(
+      borderRadius: borderRadius ?? this.borderRadius,
+      borderSide: borderSide ?? this.borderSide,
+      cut: cut ?? this.cut,
+      gapPadding: gapPadding ?? this.gapPadding,
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/cut_corners_border.dart]: Nice to keep the parameters in the same order

File: examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+import 'package:flutter/material.dart';
+import 'package:intl/intl.dart';
+import 'package:scoped_model/scoped_model.dart';
+
+import 'package:flutter_gallery/demo/shrine/model/app_state_model.dart';
+import 'package:flutter_gallery/demo/shrine/model/product.dart';
+
+class ProductCard extends StatelessWidget {
+  const ProductCard({this.imageAspectRatio = 33 / 49, this.product})
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart]:
``` 
const ProductCard({ this.imageAspectRatio = 33.0 / 49.0, this.product })
```

File: examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+      : assert(imageAspectRatio == null || imageAspectRatio > 0);
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart]: 0.0

File: examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+    final ThemeData theme = Theme.of(context);
+
+    final Image imageWidget = Image.asset(
+      product.assetName,
+      package: product.assetPackage,
+      fit: BoxFit.cover,
+    );
+
+    return ScopedModelDescendant<AppStateModel>(
+      builder: (BuildContext context, Widget child, AppStateModel model) =>
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/product_card.dart]: `=>` notation just for one liners that return a value

File: examples/flutter_gallery/lib/demo/shrine/supplemental/product_columns.dart [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return LayoutBuilder(builder: (BuildContext context, BoxConstraints constraints) {
+      const double spacerHeight = 44.0;
+
+      final double heightOfCards = (constraints.biggest.height - spacerHeight) / 2.0;
+      final double heightOfImages = heightOfCards - ProductCard.kTextBoxHeight;
+      final double imageAspectRatio = (heightOfImages >= 0.0 && constraints.biggest.width > heightOfImages)
+          ? constraints.biggest.width / heightOfImages
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/product_columns.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine/supplemental/product_columns.dart [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+          : 33 / 49;
+
+      return ListView(
+        children: <Widget>[
+          Padding(
+            padding: const EdgeInsetsDirectional.only(start: 28.0),
+            child: top != null
+                ? ProductCard(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine/supplemental/product_columns.dart]: two space indent

File: examples/flutter_gallery/lib/demo/shrine_demo.dart [hunk 28]
```diff
@@ -3,40 +3,18 @@
...
-  ShrinePageRoute({
-    WidgetBuilder builder,
-    RouteSettings settings,
-  }) : super(builder: builder, settings: settings);
-
-  @override
-  Widget buildPage(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) {
-    return buildShrine(context, super.buildPage(context, animation, secondaryAnimation));
+class ShrineDemo extends StatelessWidget {
+  ShrineDemo({Key key}) : super(key: key) {
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/shrine_demo.dart]:
```
const ShrineDemo({ Key key }) : super(key: key)
```

File: examples/flutter_gallery/lib/gallery/app.dart [hunk 29]
```diff
@@ -113,9 +117,11 @@ class _GalleryAppState extends State<GalleryApp> {
       optionsPage: GalleryOptionsPage(
         options: _options,
         onOptionsChanged: _handleOptionsChanged,
-        onSendFeedback: widget.onSendFeedback ?? () {
-          launch('https://github.com/flutter/flutter/issues/new', forceSafariVC: false);
-        },
+        onSendFeedback: widget.onSendFeedback ??
+            () {
+              launch('https://github.com/flutter/flutter/issues/new',
+                  forceSafariVC: false);
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/gallery/app.dart]: no line break (or two space indent)

File: examples/flutter_gallery/lib/gallery/app.dart [hunk 30]
```diff
@@ -113,9 +117,11 @@ class _GalleryAppState extends State<GalleryApp> {
       optionsPage: GalleryOptionsPage(
         options: _options,
         onOptionsChanged: _handleOptionsChanged,
-        onSendFeedback: widget.onSendFeedback ?? () {
-          launch('https://github.com/flutter/flutter/issues/new', forceSafariVC: false);
-        },
+        onSendFeedback: widget.onSendFeedback ??
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/gallery/app.dart]: Typically:
```
onSendFeedback: widget.onSendFeedback ?? () {
},
```

File: examples/flutter_gallery/lib/gallery/app.dart [hunk 31]
```diff
@@ -125,22 +131,25 @@ class _GalleryAppState extends State<GalleryApp> {
...
+    return ScopedModel<AppStateModel>(
+      model: model,
+      child: MaterialApp(
+        theme: _options.theme.data.copyWith(platform: _options.platform),
+        title: 'Flutter Gallery',
+        color: Colors.grey,
+        showPerformanceOverlay: _options.showPerformanceOverlay,
+        checkerboardOffscreenLayers: _options.showOffscreenLayersCheckerboard,
+        checkerboardRasterCacheImages:
+            _options.showRasterCacheImagesCheckerboard,
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/gallery/app.dart]: no line break

File: examples/flutter_gallery/lib/gallery/demos.dart [hunk 32]
```diff
@@ -93,7 +94,7 @@ List<GalleryDemo> _buildGalleryDemos() {
       icon: GalleryIcons.shrine,
       category: _kDemos,
       routeName: ShrineDemo.routeName,
-      buildRoute: (BuildContext context) => ShrineDemo(),
+      buildRoute: (BuildContext context) => ShrineApp(),
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/gallery/demos.dart]: `const ShrineApp()` (if that's possible)

**Review by HansMuller — APPROVED** [review]
I went through the code once and noted everything that wasn't typical of Formatting. I apologize for the nit pick volume. There's that and one or two meaningful revisions.

File: examples/flutter_gallery/lib/demo/shrine/app.dart [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+  // Controller to coordinate both the opening/closing of backdrop and sliding
+  // of expanding bottom sheet
+  AnimationController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = AnimationController(
+      vsync: this,
+      duration: Duration(milliseconds: 450),
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/app.dart]: Done.

File: examples/flutter_gallery/lib/demo/shrine/app.dart [hunk 34]
```diff
@@ -0,0 +1,137 @@
...
+import 'package:flutter_gallery/demo/shrine/login.dart';
+import 'package:flutter_gallery/demo/shrine/supplemental/cut_corners_border.dart';
+
+class ShrineApp extends StatefulWidget {
+  @override
+  _ShrineAppState createState() => _ShrineAppState();
+}
+
+class _ShrineAppState extends State<ShrineApp>
+    with SingleTickerProviderStateMixin {
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/app.dart]: Fixed.

File: examples/flutter_gallery/lib/demo/shrine/app.dart [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+    buttonColor: kShrinePink100,
+    scaffoldBackgroundColor: kShrineBackgroundWhite,
+    cardColor: kShrineBackgroundWhite,
+    textSelectionColor: kShrinePink100,
+    errorColor: kShrineErrorRed,
+    buttonTheme: const ButtonThemeData(
+      colorScheme: kShrineColorScheme,
+      textTheme: ButtonTextTheme.normal,
+    ),
+    primaryIconTheme: base.iconTheme.copyWith(color: kShrineBrown900),
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/app.dart]: Done.

File: examples/flutter_gallery/lib/demo/shrine/category_menu_page.dart [hunk 36]
```diff
@@ -0,0 +1,84 @@
...
+      builder: (BuildContext context, Widget child, AppStateModel model) =>
+          GestureDetector(
+            onTap: () {
+              model.setCategory(category);
+              if (onCategoryTap != null) {
+                onCategoryTap();
+              }
+            },
+            child: model.selectedCategory == category
+                ? Column(
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/category_menu_page.dart]: Done.

File: examples/flutter_gallery/lib/demo/shrine/category_menu_page.dart [hunk 37]
```diff
@@ -0,0 +1,84 @@
...
+                        height: 2.0,
+                        color: kShrinePink400,
+                      ),
+                    ],
+                  )
+                : Padding(
+                    padding: const EdgeInsets.symmetric(vertical: 16.0),
+                    child: Text(
+                      categoryString,
+                      style: theme.textTheme.body2
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/category_menu_page.dart]: Went with the latter.

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 38]
```diff
@@ -0,0 +1,658 @@
...
+  static _ExpandingBottomSheetState of(BuildContext context, {bool isNullOk = false}) {
+    assert(isNullOk != null);
+    assert(context != null);
+    final _ExpandingBottomSheetState result = context
+        .ancestorStateOfType(const TypeMatcher<_ExpandingBottomSheetState>());
+    if (isNullOk || result != null) {
+      return result;
+    }
+    throw FlutterError(
+        'ExpandingBottomSheet.of() called with a context that does not contain a ExpandingBottomSheet.\n');
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Done.

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 39]
```diff
@@ -0,0 +1,658 @@
...
+
+  final AnimationController hideController;
+
+  @override
+  _ExpandingBottomSheetState createState() => _ExpandingBottomSheetState();
+
+  static _ExpandingBottomSheetState of(BuildContext context, {bool isNullOk = false}) {
+    assert(isNullOk != null);
+    assert(context != null);
+    final _ExpandingBottomSheetState result = context
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Done.

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 40]
```diff
@@ -0,0 +1,658 @@
...
+        'ExpandingBottomSheet.of() called with a context that does not contain a ExpandingBottomSheet.\n');
+  }
+}
+
+// Emphasized Easing is a motion curve that has an organic, exciting feeling.
+// It's very fast to begin with and then very slow to finish. Unlike standard
+// curves, like [Curves.fastOutSlowIn], it can't be expressed in a cubic bezier
+// curve formula. It's quintic, not cubic. But it _can_ be expressed as one
+// curve followed by another, which we do here.
+Animation<T> _getEmphasizedEasingAnimation<T>(
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Done.

File: examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart [hunk 41]
```diff
@@ -0,0 +1,658 @@
...
+      );
+    }
+  }
+
+  Animation<double> _getThumbnailOpacityAnimation() {
+    return Tween<double>(begin: 1.0, end: 0.0).animate(
+      CurvedAnimation(
+        parent: _controller.view,
+        curve: _controller.status == AnimationStatus.forward
+            ? const Interval(0.0, 0.3)
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply examples/flutter_gallery/lib/demo/shrine/expanding_bottom_sheet.dart]: Done.

**Comment by RedBrogdon** [comment]
Obviated by #23539.

**Pull request closed without merging**